File: mergeSort.py
def merge(a1,a2,a):
    i=0
    j=0
    k=0
    while((i < len(a1)) and (j< len(a2))):
        if(a1[i]<a2[j]):
            a[k]=a1[i]
            k=k+1
            i=i+1
        else:
            a[k]=a2[j]
            k=k+1
            j=j+1
    while(i< len(a1)):
        a[k]=a1[i]
        k=k+1
        i=i+1
    while(j<len(a2)):
        a[k]=a2[j]
        k=k+1
        j=j+1                

def mergeSort(a):
    l=len(a)
    if l==0 or l==1:
        return 
    mid=l//2
    a1=a[0:mid]
    a2=a[mid:]
    mergeSort(a1)
    mergeSort(a2)
    merge(a1,a2,a)    

a=[4,5,6,4,6,21,2,4,2,5,54]
mergeSort(a)
print(a)

File: mergesort.java
import java.util.ArrayList;

class Solution
{
    void merge(int arr[], int l, int m, int r)
    {
         ArrayList<Integer> temp=new ArrayList<>();
         int left=l,right=m+1;
         while(left<=m && right<=r){
             if(arr[left]<arr[right]){
                 temp.add(arr[left]);
                 left++;
             }
             else{
                 temp.add(arr[right]);
                 right++;
             }
         }
         while(left<=m){
            temp.add(arr[left]);
            left++; 
         }
         while(right<=r){
             temp.add(arr[right]);
             right++;
         }
         //now copy contents of temp into arr  
         for(int i=l;i<=r;i++){
             arr[i]=temp.get(i-l);//this part is amazing
         }
    }
    void mergeSort(int arr[], int l, int r)
    {
        if(l>=r)
        return;
        int mid=l+(r-l)/2;
        mergeSort(arr,l,mid);
        mergeSort(arr,mid+1,r);
        merge(arr,l,mid,r);
    }
}